Source/DafnyCore/Verifier/ProofObligationDescription.cs

```diff
@@ -308,6 +308,45 @@ public class AssertStatement : ProofObligationDescription {
   }
 }
 
+// The Boogie version does not support custom error messages yet
+public class RequiresDescription : ProofObligationDescription {
+  private readonly string customErrMsg;
+  public override string SuccessDescription =>
+    customErrMsg is null
+      ? "the precondition always holds"
+      : $"error is impossible: {customErrMsg}";
+
+  public override string FailureDescription =>
+    customErrMsg ?? "a precondition could not be proven";
+
+  public override string ShortDescription => "requires";
+
+  public RequiresDescription([CanBeNull] string customErrMsg = null) {
+    this.customErrMsg = customErrMsg;
+  }
+}
+
+// The Boogie version does not support custom error messages yet
+public class EnsuresDescription : ProofObligationDescription {
+  private readonly string customErrMsg;
+  public override string SuccessDescription =>
+    customErrMsg is null
+      ? "this postcondition holds"
+      : $"error is impossible: {customErrMsg}";
```

Comment:
Why not take a `customSuccessMessage` as well? `error is impossible:` sounds bad.

Reply:
> Why not take a customSuccessMessage as well? error is impossible: sounds bad.

It took me some time to implement but I thought it was a nice idea.
I implemented (and documented it) using the same attribute `{:error <error message>[, <success message>`}` and I dropped the string "error is impossible" entirely. I also refactored the code so that similar proof obligation descriptions use the same overriding mechanism.